controller/controller.go

```diff
@@ -1651,6 +1676,26 @@ func getInClusterNamespace() string {
 	return "default"
 }
 
+func (ctrl *ProvisionController) makeVolumeName(pvcUID string) (string, error) {
+	// create persistent name based on a volumeNamePrefix and volumeNameUUIDLength
+	// of PVC's UID
+	if len(ctrl.volumeNamePrefix) == 0 {
+		return "", fmt.Errorf("Volume name prefix cannot be of length 0")
+	}
+	if ctrl.volumeNameUUIDLength == 0 {
+		return "", fmt.Errorf("Volume length cannot be 0")
+	}
+	if len(pvcUID) == 0 {
+		return "", fmt.Errorf("corrupted PVC object, it is missing UID")
+	}
+	if ctrl.volumeNameUUIDLength == -1 {
+		// Default behavior is to not truncate or remove dashes
+		return fmt.Sprintf("%s-%s", ctrl.volumeNamePrefix, pvcUID), nil
+	}
+	// Else we remove all dashes from UUID and truncate to volumeNameUUIDLength
+	return fmt.Sprintf("%s-%s", ctrl.volumeNamePrefix, strings.Replace(string(pvcUID), "-", "", -1)[0:ctrl.volumeNameUUIDLength]), nil
+}
```

Comment:
I think it would be better to keep this weird function in the external-provisioner and here just provide a configurable callback stored in `ProvisionController`:

```go

type VolumeNameHook func(claim *v1.PersistentVolumeClaim) string
type ProvisionController struct {
    ...
    volumeNameHook VolumeNameHook
}
...

func VolumeName(hook VolumeNameHook) func(*ProvisionController) error {
	return func(c *ProvisionController) error {
		if c.HasRun() {
			return errRuntime
		}
		c.volumeNameHook = hook
		return nil
	}
}
```

With the default value `getProvisionedVolumeNameForClaim`.

Reply:
Thank you for your suggestion. I try to do this.

Reply:
Done. Please check @jsafrane